src/main/java/fr/moribus/imageonmap/image/MapInitEvent.java

```diff
@@ -131,4 +146,115 @@ public void onPlayerInventoryPlace(InventoryClickEvent event) {
+                switch (event.getCause()) {
+                    case MAGIC:
+                    case ENTITY_EXPLOSION:
+                    case FIRE_TICK:
+                    case LIGHTNING:
+                    case CRAMMING:
+                    case WITHER:
+                    case SUFFOCATION:
+                    case DROWNING:
+                    case BLOCK_EXPLOSION:
+                        event.setCancelled(true);
+                        break;
+                    default:
+                }
+            }
+        }
+    }
+
+
+    @EventHandler
+    public void onHangingBreakEvent(HangingBreakEvent event) {
+        Entity entity = event.getEntity();
+        PluginLogger.info("entity " + entity.toString());
+        if (!(entity instanceof ItemFrame)) {
+            return;
+        }
+
+        ItemStack item = ((ItemFrame) entity).getItem();
+        if (item.getType() == Material.FILLED_MAP) {
+            //if the map exist we canceled the event
+            if (MapManager.getMap(item) != null) {
+                if (event.getCause() == HangingBreakEvent.RemoveCause.EXPLOSION) {
+                    //creeper goes boom
+                    event.setCancelled(true);
+                }
+            }
+        }
+    }
+
+
+    protected static final class EntitiesLoadListener implements Listener {
+        @FutureEventHandler(event = "world.EntitiesLoadEvent")
+        public void onEntitiesLoad(WrappedEvent event) {
+            //New in 1.17
+            //Used to make sure map are really loaded in 1.17 on Paper (else some won't render or update properly)
+            RunTask.later(() -> {
+                try {
+                    Chunk chunk = (Chunk) Reflection.call(event.getEvent(), "getChunk");
+                    Entity[] entities = chunk.getEntities();
+                    //Not the most efficient method because we go through entity already loaded
+
+                    //The direct method using getEntities of
+                    // https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/world/EntitiesLoadEvent.html
+                    //Return an unmodifiable list of entities.
+                    //Using this make the overall process a bit more efficient but way more complicated and very weak
+                    // to change.
+                    //TODO Investigate if there is a better way to do this.
+                    //Early exit, most are empty entities array
+                    if (entities.length == 0) {
+                        return;
+                    }
+
+                    for (Entity entity : entities) {
+                        if (entity instanceof ItemFrame) {
+                            initMap(((ItemFrame) entity).getItem());
+                        }
+                    }
+                } catch (Exception e) {
+                    PluginLogger.error(e.toString());
+                    return;
+                }
+
+            }, 5L);
+        }
```

Comment:
No problems with this (way less complicated) version:

```java
@FutureEventHandler(event = "world.EntitiesLoadEvent")
public void onEntitiesLoad(WrappedEvent wrappedEvent) {
    EntitiesLoadEvent event = (EntitiesLoadEvent) wrappedEvent.getEvent();
    for (Entity entity : event.getEntities()) {
        if (entity instanceof ItemFrame) {
            initMap(((ItemFrame) entity).getItem());
        }
    }
}
```

Reply:
As stated in the comment that doesn't work in our case because event.getEntities() (only for EntitiesLoadEvent) return an unmodifiable list and I can't access the list items to edit them. https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/world/EntitiesLoadEvent.html#getEntities That's why I go through this workaround. :/